src/api/artifact/controller.go

```diff
@@ -347,67 +347,68 @@ func (c *controller) deleteDeeply(ctx context.Context, id int64, isRoot bool) er
 }
 
 func (c *controller) Copy(ctx context.Context, srcRepo, reference, dstRepo string) (int64, error) {
-	return c.copyDeeply(ctx, srcRepo, reference, dstRepo, true)
-}
-
-// as we call the docker registry APIs in the registry client directly,
-// this bypass our own logic(ensure, fire event, etc.) inside the registry handlers,
-// these logic must be covered explicitly here.
-// "copyDeeply" iterates the child artifacts and copy them first
-func (c *controller) copyDeeply(ctx context.Context, srcRepo, reference, dstRepo string, isRoot bool) (int64, error) {
-	var option *Option
-	// only get the tags of the root parent
-	if isRoot {
-		option = &Option{WithTag: true}
-	}
-
-	srcArt, err := c.GetByReference(ctx, srcRepo, reference, option)
+	option := &Option{WithTag: true}
+	root, err := c.GetByReference(ctx, srcRepo, reference, option)
 	if err != nil {
 		return 0, err
 	}
 
-	digest := srcArt.Digest
+	// Copy to self, skip it
+	if srcRepo == dstRepo {
+		return root.ID, nil
+	}
 
-	// check the existence of artifact in the destination repository
-	dstArt, err := c.GetByReference(ctx, dstRepo, digest, option)
-	if err == nil {
-		// return conflict error if the root parent artifact already exists under the destination repository
-		if isRoot {
-			return 0, ierror.New(nil).WithCode(ierror.ConflictCode).
-				WithMessage("the artifact %s@%s already exists", dstRepo, digest)
+	var copyArtifacts []*Artifact
+	walkFn := func(a *Artifact) error {
+		digest := a.Digest
+
+		_, err := c.GetByReference(ctx, dstRepo, digest, nil)
```

Comment:
what if the source artifact have tags but doesn't have in target? The tags are lost?

I think we can leverge Ensure to judge whether to create in the target.